File: day09/day09.zig
const std = @import("std");
const print = std.debug.print;
const ArrayList = std.ArrayList;

const Point = struct {
    x: i32,
    y: i32,

    const Self = @This();

    pub fn new() Self {
        return .{ .x = 0, .y = 0 };
    }

    pub fn newXY(x: i32, y: i32) Self {
        return .{ .x = x, .y = y };
    }
};

pub fn syncPosition(head: *Point, tail: *Point) void {
    const dx = head.x - tail.x;
    const dy = head.y - tail.y;
    if (std.math.absCast(dx) > 1 or std.math.absCast(dy) > 1) {
        tail.x += std.math.sign(dx);
        tail.y += std.math.sign(dy);
    }
}

const MOVE_MAP = std.ComptimeStringMap(Point, .{
    .{ "U", Point.newXY(0, 1) },
    .{ "D", Point.newXY(0, -1) },
    .{ "R", Point.newXY(1, 0) },
    .{ "L", Point.newXY(-1, 0) },
});

pub fn solve(allocator: std.mem.Allocator, content: []const u8, comptime ropeLen: usize) !u32 {
    var knots: [ropeLen]Point = .{Point.new()} ** ropeLen;
    var lineIter = std.mem.tokenize(u8, content, "\n");
    var tailTrace = std.AutoHashMap(Point, bool).init(allocator);
    while (lineIter.next()) |line| {
        var cmdIter = std.mem.tokenize(u8, line, " ");
        const direction = MOVE_MAP.get(cmdIter.next().?).?;
        const steps = try std.fmt.parseInt(i32, cmdIter.next().?, 10);
        var count: usize = 0;
        while (count < steps) {
            knots[0].x += direction.x;
            knots[0].y += direction.y;
            var knotId: usize = 1;
            while (knotId < ropeLen) {
                syncPosition(&knots[knotId - 1], &knots[knotId]);
                knotId += 1;
            }
            try tailTrace.put(knots[ropeLen - 1], true);
            count += 1;
        }
    }
    return tailTrace.count();
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const content = @embedFile("input.txt");
    const part1 = try solve(allocator, content, 2);
    const part2 = try solve(allocator, content, 10);
    print("Part 1 = {}\nPart 2 = {}", .{ part1, part2 });
}

File: day09/day09.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

var MOVE_MAP = map[string]Point{
	"R": NewPointXY(1, 0),
	"L": NewPointXY(-1, 0),
	"U": NewPointXY(0, 1),
	"D": NewPointXY(0, -1),
}

func abs(x int) int {
	if x < 0 {
		return -x
	} else {
		return x
	}
}

func sign(x int) int {
	if x < 0 {
		return -1
	}
	if x > 0 {
		return 1
	}
	return 0
}

type Point struct {
	x int
	y int
}

func NewPoint() Point {
	return Point{0, 0}
}

func NewPointXY(x, y int) Point {
	return Point{x, y}
}

func SyncPosition(head *Point, tail *Point) {
	dx := head.x - tail.x
	dy := head.y - tail.y
	if abs(dx) > 1 || abs(dy) > 1 {
		tail.x += sign(dx)
		tail.y += sign(dy)
	}
}

func Solve(lines []string, ropeLen int) int {
	knots := make([]Point, ropeLen)
	tailTrace := make(map[Point]bool)
	tailTrace[knots[1]] = true
	for _, line := range lines {
		parsed := strings.Split(line, " ")
		direction := parsed[0]
		steps, _ := strconv.ParseInt(parsed[1], 10, 0)
		delta := MOVE_MAP[direction]
		count := 0
		for count < int(steps) {
			knots[0].x += delta.x
			knots[0].y += delta.y
			knotIdx := 1
			for knotIdx < ropeLen {
				SyncPosition(&knots[knotIdx-1], &knots[knotIdx])
				knotIdx++
			}
			tailTrace[knots[ropeLen-1]] = true
			count++
		}
	}
	return len(tailTrace)
}

func main() {
	content, _ := os.ReadFile("input.txt")
	lines := strings.Split(strings.Trim(string(content), "\n"), "\n")
	fmt.Printf("Part 1 = %d\n", Solve(lines, 2))
	fmt.Printf("Part 2 = %d\n", Solve(lines, 10))
}
